langchain/callbacks/arthur_callback.py

```diff
@@ -0,0 +1,198 @@
+    ) -> None:
+        """Initialize callback handler."""
+
+        super().__init__()
+        from arthurai import ArthurAI
+        from arthurai.common.constants import InputType, OutputType, Stage, ValueType
+        from arthurai.common.exceptions import ResponseClientError
+        from arthurai.util import generate_timestamps
+        
+        # save the Arthur timestamp function to be used to create valid inference timestamps in on_llm_end()
+        self.timestamp_fn = generate_timestamps
+                
+        # connect to Arthur               
+        if arthur_login is None:
+            try:
+                arthur_api_key = os.environ['ARTHUR_API_KEY']
+            except KeyError:
+                raise ValueError('No Arthur authentication provided. Either give a login to the ArthurCallbackHandler \
+                    or set an ARTHUR_API_KEY as an environment variable.')
+            arthur = ArthurAI(url=arthur_url, access_key=arthur_api_key)
+        else:
+            if arthur_password is None:
+                arthur = ArthurAI(url=arthur_url, login=arthur_login)
+            else:
+                arthur = ArthurAI(url=arthur_url, login=arthur_login, password=arthur_password)
+                
+        # get model from Arthur by the provided model ID
+        try:
+            self.arthur_model = arthur.get_model(model_id)
+        except ResponseClientError:
+            raise ValueError(f"Was unable to retrieve model with id {model_id} from Arthur. \
+                Make sure the ID corresponds to a model that is currently registered with your Arthur account.")
+            
+        # save the attributes of this model to be used when preparing inferences to log to Arthur in on_llm_end()
+        self.attr_names = set([a.name for a in self.arthur_model.get_attributes()])
+        self.input_attr = [x for x in self.arthur_model.get_attributes()
+                           if x.stage==Stage.ModelPipelineInput and x.value_type==ValueType.Unstructured_Text][0].name
+        self.output_attr = [x for x in self.arthur_model.get_attributes()
+                           if x.stage==Stage.PredictedValue and x.value_type==ValueType.Unstructured_Text][0].name
+        self.token_likelihood_attr = None
+        if len([x for x in self.arthur_model.get_attributes() if x.value_type==ValueType.TokenLikelihoods]) > 0:
+            self.token_likelihood_attr = [x for x in self.arthur_model.get_attributes()
+                           if x.value_type==ValueType.TokenLikelihoods][0].name
+        
+        # prepare callback data defaults to be updated in on_llm_start()
+        self.input_texts: List = []
+        self.on_llm_start_time: float = -1.0
+        
+
+    def on_llm_start(
+        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
+    ) -> None:
+        """On LLM start, save the input prompts"""
+        self.input_texts = prompts
+        self.on_llm_start_time = time()
```

Comment:
Oh missed  pretty big one -> callback handlers can be called asynchronously, which means one LLM could start before the previous LLM ends, breaking this state tracking as a single variable .

All the `on_llm_?` methods are passed a `run_id: UUID` keyword argument you can use to make sure the state you are tracking is associated with the correct run.

You will probably want to do something like 

```
self.run_map = defaultdict(dict)
```

and then

```
self.run_map[run_id]['on_llm_start_time'] = time()
self.run_map[run_id]['input_texts'] = prompts
...
time_from_start_to_end = self.run_map[run_id]['on_llm_start_time'] -time()
```

adding in the appropriate error handling in case the run id isn't present in the dicitionary.